File: Javascript/selection_sort.js
//Javascript implementation of selection sort

//recieves an array of unsorted integers, find's the smallest integer,
//and return it's index
const findSmallest = arr => {
  let smallest = arr[0];
  //postion of the smallest input array
  let position = 0;
  //iterate through input array and find the smallest integer
  arr.map((value, index) => {
    if (value < smallest) {
      smallest = value;
      position = index;
    }
  });
  return position;
};

//recieves an array of unsorted integer and return it sorted
const selectionSort = arr => {
  //create a copy of the input array, to prevent weird behaviours
  //when we modify the input array
  let temparr = arr.slice(0);
  //new aray for storing sorted items
  let newArr = [];

  for (let _i of arr) {
    //store index of smallest integer returned by the findSmallest function
    let index = findSmallest(temparr);
    //remove smallest interger from tempArr and push to newArr
    newArr.push(...temparr.splice(index, 1));
  }
  return newArr;
};

//Tests
let arr = [];
let temp = 10000;
while (temp > 0) {
  arr.push(temp);
  temp = temp - 1;
}
console.time('sort');
console.log(selectionSort(arr));
console.timeEnd('sort');

File: Python/selection_sort.py
#Implementation of selection sort in python

#recieves an array of unsorted integers
#and returns the index of the smallest integer
def find_smallest(arr):
    smallest = arr[0];
    postion = 0;
    # iterate through array and return index of the smallest integer
    for i in range(0,len(arr)):
        if arr[i] < smallest:
            smallest = arr[i]
            postion = i

    return postion


def selection_sort(arr):
    #python makes use of reference counting so directly modifying
    #the array would give me issues so i create a copy of the input array
    new_array = arr[0:len(arr)]
    sorted_array = []

    for i in arr:
        index = find_smallest(new_array)
        #remove smallest integer from array and push to sorted array
        sorted_array.append(new_array.pop(index))

    return sorted_array

#Tests
temp = 10000
arr = []
while temp > 0:
    arr.append(temp)
    temp = temp - 1

print(selection_sort(arr))